File: library-common/src/main/java/library/common/framework/ui/activity/view/ViewController.java
package library.common.framework.ui.activity.view;

import android.app.Activity;
import android.content.res.Resources;
import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.StringRes;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import butterknife.ButterKnife;

/**
 * 子View视图控制器，用来更加细颗粒的拆分{@link AppDelegate}
 *
 * @author hiphonezhu@gmail.com
 * @version [AndroidLibrary, 2019-1-23]
 */
public abstract class ViewController {
    protected AppDelegate mAppDelegate;
    View mView;

    public ViewController(@NonNull AppDelegate appDelegate, int parentId) {
        mAppDelegate = appDelegate;
        ViewGroup parent = appDelegate.getContentView().findViewById(parentId);
        mView = LayoutInflater.from(parent.getContext()).inflate(getView(), parent, true);
        ButterKnife.bind(this, parent);
        appDelegate.addViewController(this);
        initWidget();
    }

    /**
     * 返回View布局
     *
     * @return
     */
    public abstract @LayoutRes
    int getView();

    public void setVisibility(int visibility) {
        mView.setVisibility(visibility);
    }

    public void setOnClickListener(final View.OnClickListener listener, int... ids) {
        mAppDelegate.setOnClickListener(listener, ids);
    }

    public boolean isVisible() {
        return mAppDelegate.isVisible();
    }

    protected void initWidget() {
    }

    protected <T extends Activity> T getActivity() {
        return mAppDelegate.getActivity();
    }

    protected <T extends Fragment> T getFragment() {
        return mAppDelegate.getFragment();
    }

    protected Resources getResources() {
        return mAppDelegate.getResources();
    }

    protected String getString(@StringRes int strRes) {
        return mAppDelegate.getString(strRes);
    }

    protected String getString(@StringRes int strRes, Object... formatArgs) {
        return mAppDelegate.getString(strRes, formatArgs);
    }

    protected void showToast(CharSequence message) {
        mAppDelegate.showToast(message);
    }

    protected void showProgress(String message, boolean cancelable) {
        mAppDelegate.showProgress(message, cancelable);
    }

    protected void hideProgress() {
        mAppDelegate.hideProgress();
    }

    protected void onShow() {

    }

    protected void onHide() {

    }

    protected void onDestroy() {

    }
}

File: library_android/src/main/java/library/common/framework/ui/activity/view/ViewController.kt
package library.common.framework.ui.activity.view

import android.app.Activity
import android.content.res.Resources
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.annotation.StringRes
import androidx.fragment.app.Fragment

/**
 * 子 View 视图控制器，用来更加细颗粒的拆分 [AppDelegate]
 *
 * @author hiphonezhu@gmail.com
 * @version [AndroidLibrary, 2019-1-23]
 */
abstract class ViewController(private val appDelegate: AppDelegate, parentId: Int) {

    val view: View

    init {
        val parent = appDelegate.contentView.findViewById<ViewGroup>(parentId)
        view = LayoutInflater.from(parent.context).inflate(getView(), null, false)
        parent.removeAllViews()
        parent.addView(view)
        appDelegate.addViewController(this)
    }

    var viewBindingObj: Any? = null

    inline fun <reified T> getViewBinding(): T {
        return (viewBindingObj as T?) ?: let {
            val bindMethod = T::class.java.getMethod("bind", View::class.java)
            (bindMethod.invoke(null, view) as T).apply {
                viewBindingObj = this
            }
        }
    }

    open fun initWidget() {}

    fun onShow() {}

    fun onHide() {}

    fun onDestroy() {}

    /**
     * 返回View布局
     *
     * @return
     */
    @LayoutRes
    abstract fun getView(): Int

    protected fun setOnClickListener(listener: View.OnClickListener?, vararg ids: Int) {
        appDelegate.setOnClickListener(listener, *ids)
    }

    protected fun <T : Activity?> getActivity(): T {
        return appDelegate.getActivity()
    }

    protected fun <T : Fragment?> getFragment(): T? {
        return appDelegate.getFragment()
    }

    protected fun getResource(): Resources = appDelegate.getResources()

    protected fun getString(@StringRes strRes: Int): String {
        return appDelegate.getString(strRes)
    }

    protected fun getString(@StringRes strRes: Int, vararg formatArgs: Any?): String {
        return appDelegate.getString(strRes, *formatArgs)
    }

    protected fun showToast(message: String) {
        appDelegate.showToast(message)
    }
}